src/Microsoft.TestPlatform.Common/Filtering/FastFilter.cs

```diff
@@ -0,0 +1,161 @@
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.VisualStudio.TestPlatform.Common.Filtering
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Collections.Immutable;
+    using System.Linq;
+    using System.Text.RegularExpressions;
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel;
+
+    internal sealed class FastFilter
+    {
+        internal string FilterPropertyName { get; }
+
+        internal ImmutableHashSet<string> FilterPropertyValues { get; }
+
+        internal bool IsFilteredOutWhenMatched { get; }
+
+        internal Regex PropertyValueRegex { get; set; }
+
+        internal FastFilter(string filterPropertyName, ImmutableHashSet<string> filterPropertyValues, Operation filterOperation, Operator filterOperator)
+        {
+            ValidateArg.NotNullOrEmpty(filterPropertyName, nameof(filterPropertyName));
+            ValidateArg.NotNull(filterPropertyValues, nameof(filterPropertyValues));
+            
+            this.FilterPropertyName = filterPropertyName;
+            this.FilterPropertyValues = filterPropertyValues;
+
+            if (filterOperation == Operation.Equal && (filterOperator == Operator.Or || filterOperator == Operator.None))
+            {
+                IsFilteredOutWhenMatched = false;
+            }
+            else if (filterOperation == Operation.NotEqual && (filterOperator == Operator.And || filterOperator == Operator.None))
+            {
+                IsFilteredOutWhenMatched = true;
+            }
+            else
+            {
+                throw new ArgumentException();
+            }
+        }
+
+        internal string[] ValidForProperties(IEnumerable<string> properties)
+        {
+            return properties.Contains(this.FilterPropertyName, StringComparer.OrdinalIgnoreCase)
+                ? null
+                : new[] { this.FilterPropertyName };
+        }
+
+        internal bool Evaluate(Func<string, Object> propertyValueProvider)
+        {
+            ValidateArg.NotNull(propertyValueProvider, "propertyValueProvider");
+
+            if (!TryGetSinglePropertyValue(this.FilterPropertyName, propertyValueProvider, out var value))
+            {
+                return false;
+            }
+
+            if (PropertyValueRegex != null)
+            {
+                var match = PropertyValueRegex.Match(value);
+                if (match.Success)
+                {
+                    value = match.Value;
+                }
+                else
+                {
+                    return false;
+                }
+            }
+
+            var matched = this.FilterPropertyValues.Contains(value);
+            return IsFilteredOutWhenMatched ? !matched : matched;
+        }
+
+        private static bool TryGetSinglePropertyValue(string name, Func<string, Object> propertyValueProvider, out string singleValue)
```

Comment:
nit: Looks like there is only one caller. Can we inline this?

Reply:
Sure.